drivers/gpio/gpio_max14906.c

```diff
@@ -0,0 +1,495 @@
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(gpio_max14906);
+
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+#include "gpio_max14906.h"
+#include "gpio_max149x6.h"
+
+#define DT_DRV_COMPAT adi_max14906_gpio
+
+static int gpio_max14906_diag_chan_get(const struct device *dev);
+
+static int max14906_pars_spi_diag(const struct device *dev, uint8_t *rx_diag_buff, uint8_t rw)
+{
+	struct max14906_data *data = dev->data;
+	int ret = 0;
+
+	if (rx_diag_buff[0]) {
+		LOG_ERR("[DIAG] MAX14906 in SPI diag - error detected\n");
+		data->glob.interrupt.reg_bits.SHT_VDD_FAULT = GET_BIT(rx_diag_buff[0], 5);
+		data->glob.interrupt.reg_bits.ABOVE_VDD_FAULT = GET_BIT(rx_diag_buff[0], 4);
+		data->glob.interrupt.reg_bits.OW_OFF_FAULT = GET_BIT(rx_diag_buff[0], 3);
+		data->glob.interrupt.reg_bits.CURR_LIM = GET_BIT(rx_diag_buff[0], 2);
+		data->glob.interrupt.reg_bits.OVER_LD_FAULT = GET_BIT(rx_diag_buff[0], 1);
+
+		uint8_t globlf = GET_BIT(rx_diag_buff[0], 0);
+
+		ret = -EIO;
+
+		PRINT_ERR(data->glob.interrupt.reg_bits.SHT_VDD_FAULT);
+		PRINT_ERR(data->glob.interrupt.reg_bits.ABOVE_VDD_FAULT);
+		PRINT_ERR(data->glob.interrupt.reg_bits.OW_OFF_FAULT);
+		PRINT_ERR(data->glob.interrupt.reg_bits.CURR_LIM);
+		PRINT_ERR(data->glob.interrupt.reg_bits.OVER_LD_FAULT);
+		PRINT_ERR(globlf);
+	}
+
+	if (rw == MAX149x6_WRITE && (rx_diag_buff[1] & 0x0f)) {
+		/* +-----------------------------------------------------------------------+
+		 * | LSB                             BYTE 2                            MSB |
+		 * +--------+--------+--------+--------+--------+--------+--------+--------+
+		 * |   BIT0 |   BIT1 |   BIT2 |   BIT3 |   BIT4 |   BIT5 |   BIT6 |   BIT7 |
+		 * +--------+--------+--------+--------+--------+--------+--------+--------+
+		 * | Fault1 | Fault2 | Fault3 | Fault4 | DiLvl1 | DiLvl2 | DiLvl3 | DiLvl4 |
+		 * +--------+--------+--------+--------+--------+--------+--------+--------+
+		 */
+
+		LOG_ERR("[DIAG] Flt1[%x] Flt2[%x] Flt3[%x] Flt4[%x]", GET_BIT(rx_diag_buff[1], 0),
+								      GET_BIT(rx_diag_buff[1], 1),
+								      GET_BIT(rx_diag_buff[1], 2),
+								      GET_BIT(rx_diag_buff[1], 3));
+		if (rx_diag_buff[1] & 0x0f) {
+			LOG_ERR("[DIAG] gpio_max14906_diag_chan_get(%x)\n", rx_diag_buff[1] & 0x0f);
+			ret = gpio_max14906_diag_chan_get(dev);
+		}
+	}
+
+	return ret;
+}
+
+static int max14906_reg_trans_spi_diag(const struct device *dev, uint8_t addr,
+					uint8_t tx, uint8_t rw)
+{
+	const struct max14906_config *config = dev->config;
+	uint8_t rx_diag_buff[2];
+
+	if (!gpio_pin_get_dt(&config->fault_gpio)) {
+		LOG_ERR("[FAULT] pin triggered");
+	}
+
+	uint8_t ret = max149x6_reg_transceive(dev, addr, tx, rx_diag_buff, rw);
+
+	if (max14906_pars_spi_diag(dev, rx_diag_buff, rw)) {
+		ret = -EIO;
+	}
+
+	return ret;
+}
+
+#define max14906_reg_read(dev, addr) max14906_reg_trans_spi_diag(dev, addr, 0, MAX149x6_READ)
+#define max14906_reg_write(dev, addr, val) max14906_reg_trans_spi_diag(dev, addr, val, \
```

Comment:
It is fine for me but might be not match with naming convention for macro
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-3-macro-name-collisions

Reply:
Seems macros defined with upper case (MIN, MAX, ARRAY_SIZE...) that make it more readable, when looking the code it clearly mention either it is macro or not when reading source code.

Reply:
Ok yes, that make sens. I thought you are talking about how macros names are worded.
Will replace them with upper case.